src/compiler/checker.ts

```diff
@@ -4716,19 +4731,33 @@ namespace ts {
             return result;
         }
 
+        function cloneTypePredicate(predicate: TypePredicate, mapper: TypeMapper): ThisTypePredicate | IdentifierTypePredicate {
+            if (predicate.kind === TypePredicateKind.Identifier) {

```

Comment:
I think a lot of this special casing could be eliminated by using `parameterName = "this"` and `parameterIndex = -1` to indicate a `this` type predicate.